### PR TITLE
feat: gate provider implementation

### PR DESCRIPTION
**Pull request by adamewozniak**
## Description

Adds gate.io as a provider to support [umee/usdt](https://www.gate.io/trade/UMEE_USDT)

closes: #604

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: price-feeder/oracle/provider/gate.go [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+	}
+
+	return nil
+}
+
+// subscribeCandles subscribes to the candle channels for all pairs one-by-one.
+// The gate API currently only supports subscribing to one kline market at a time.
+//
+// REF: https://www.gate.io/docs/websocket/#kline-subscription
+func (p *GateProvider) subscribeCandles(cps ...types.CurrencyPair) error {
```
Comment (adamewozniak) [inline price-feeder/oracle/provider/gate.go:207]: Just want to highlight this. Per gate's docs : 

```
Can only subscribe to one market at the same time, market list is not supported currently. 
For multiple subscriptions, only the last one takes effect.
```

tldr; we can only use gate for **one** currency pair (which will have to be umee) until they support multiple. 
We could do some strict config validation for this, but seems like we should just add a log warning. wdyt @alexanderbez ?

File: price-feeder/oracle/provider/gate_test.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+		require.Equal(t, "gate provider failed to get ticker price for FOO_BAR", err.Error())
+		require.Nil(t, prices)
+	})
+}
+
+func TestGateCurrencyPairToGatePair(t *testing.T) {
+	cp := types.CurrencyPair{Base: "ATOM", Quote: "USDT"}
+	GateSymbol := currencyPairToGatePair(cp)
+	require.Equal(t, GateSymbol, "ATOM_USDT")
+}
```
Comment (adamewozniak) [inline price-feeder/oracle/provider/gate_test.go:80]: Should definitely add tests for GetCandlePrices in a separate pr

File: price-feeder/oracle/provider/gate.go (hunk 1 above)
Comment (alexanderbez) [reply price-feeder/oracle/provider/gate.go]: lol what a shitty API. My proposal to actually add a (temporary) error to the config validation where an error is thrown if gate is used for more than one pair. WDYT?

Comment (adamewozniak) [reply price-feeder/oracle/provider/gate.go]: Yup! Agreed

**Review by alexanderbez — APPROVED** [review]
LGTM. Let's get that config change in.

**Comment by adamewozniak** [comment]
> LGTM. Let's get that config change in.

Done at ed837f455611348519a22acb5e9335ca861b3d0e

@RafilxTenfen do you want to check this out before it gets merged in?

File: price-feeder/oracle/provider/gate.go [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+	// GateTickerResponse defines the response body for gate tickers.
+	// The Params response is a 2D slice of multiple candles and their data.
+	//
+	// REF: https://www.gate.io/docs/websocket/index.html
+	GateCandleResponse struct {
+		Method string          `json:"method"`
+		Params [][]interface{} `json:"params"`
+	}
+
+	// GateEvent defines the response body for gate subscription statuses
```
Comment (RafilxTenfen) [inline price-feeder/oracle/provider/gate.go:83]:
```suggestion
	// GateEvent defines the response body for gate subscription statuses.
```

File: price-feeder/oracle/provider/gate.go [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+	GateEvent struct {
+		ID     int             `json:"id"`     // subscription id, ex.: 123
+		Result GateEventResult `json:"result"` // event result body
+	}
+	GateEventResult struct {
```
Comment (RafilxTenfen) [inline price-feeder/oracle/provider/gate.go:88]: nit: add godoc

File: price-feeder/oracle/provider/gate.go [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+	}
+	wsConn.SetPongHandler(p.pongHandler)
+	p.wsClient = wsConn
+
+	cps := make([]types.CurrencyPair, len(p.subscribedPairs))
+	iterator := 0
+	for _, cp := range p.subscribedPairs {
+		cps[iterator] = cp
+		iterator++
+	}
```
Comment (RafilxTenfen) [inline price-feeder/oracle/provider/gate.go:479]: - You could do the same as we do in the [binance](https://github.com/umee-network/umee/blob/d75ab49759e8bcdefb088d54d8cc12823e31d954/price-feeder/oracle/provider/binance.go#L320-L321)
```suggestion
	p.mtx.RLock()
	defer p.mtx.RUnlock()
	cps := mapPairsToSlice(p.subscribedPairs)
```

- or just add the mutex Read lock and unlock to it

File: price-feeder/oracle/provider/gate_test.go (hunk 2 above)
Comment (RafilxTenfen) [reply price-feeder/oracle/provider/gate_test.go:80]: agreed

**Review by RafilxTenfen — CHANGES_REQUESTED** [review]
- The only change actually needed is the mutex lock and unlock at [reconnect](https://github.com/umee-network/umee/blob/585d591d6cfa5cb579c96ff3f4cf773d29254418/price-feeder/oracle/provider/gate.go#L462-L463)

**Review by RafilxTenfen — APPROVED** [review]
